File: src/services/projects.ts
import type { Project, ServiceResult } from "@/domain/types";
import { tauriClient } from "@/services/api/tauri-client";

export const createProject = async (name: string, description: string | null) => {
  try {
    const project = await tauriClient.createProject(name, description);
    return { ok: true, data: project } satisfies ServiceResult<Project>;
  } catch (error) {
    return {
      ok: false,
      error: error instanceof Error ? error.message : "Unable to create project.",
    } satisfies ServiceResult<Project>;
  }
};

export const updateProject = async (
  id: string,
  name: string,
  description: string | null,
) => {
  try {
    const project = await tauriClient.updateProject(id, name, description);
    return { ok: true, data: project } satisfies ServiceResult<Project>;
  } catch (error) {
    return {
      ok: false,
      error: error instanceof Error ? error.message : "Unable to update project.",
    } satisfies ServiceResult<Project>;
  }
};

export const deleteProject = async (id: string) => {
  try {
    await tauriClient.deleteProject(id);
    return { ok: true } satisfies ServiceResult<void>;
  } catch (error) {
    return {
      ok: false,
      error: error instanceof Error ? error.message : "Unable to delete project.",
    } satisfies ServiceResult<void>;
  }
};

File: src-tauri/src/commands/projects.rs
use tauri::State;

use crate::{errors::app_error::AppResult, models::Project, state::app_state::AppState};

#[tauri::command]
pub async fn list_projects(state: State<'_, AppState>) -> AppResult<Vec<Project>> {
    state.project_service.list().await
}

#[tauri::command]
pub async fn create_project(
    name: String,
    description: Option<String>,
    state: State<'_, AppState>,
) -> AppResult<Project> {
    state.project_service.create(&name, description).await
}

#[tauri::command]
pub async fn update_project(
    id: String,
    name: String,
    description: Option<String>,
    state: State<'_, AppState>,
) -> AppResult<Project> {
    state.project_service.update(&id, &name, description).await
}

#[tauri::command]
pub async fn delete_project(id: String, state: State<'_, AppState>) -> AppResult<()> {
    state.project_service.delete(&id).await
}
